retry on {Node,Store}ID allocation

tbg opened the pull request:
The calls to generate Node and Store IDs should be retried when they fail.

Otherwise, Fatals like this can happen:

> node connected via gossip and verified as part of cluster "..."
> unable to allocate node ID: the descriptor for the first range is not available via gossip

tamird commented:
Can you add a test for this?

tamird commented:
LGTM but needs a test

bdarnell commented on server/node.go:
```diff
@@ -80,24 +89,50 @@ type Node struct {
+	})
 	if err != nil {
-		return 0, util.Errorf("unable to allocate node ID: %s", err)
+		log.Fatal(err)
```
Why did you move this? We already had errors being passed up the stack, so it seems better to continue to do that than to add a log.Fatal at a lower level.

tschottdorf replied:
You're right. Moved it back up.

bdarnell commented:
LGTM

tbg committed:
retry on {Node,Store}ID allocation

> node connected via gossip and verified as part of cluster "..."
> unable to allocate node ID: the descriptor for the first range is not available via gossip

tamird commented:
LGTM thanks for adding the test

tschottdorf commented:
you're welcome.

spencerkimball commented:
I think this is the wrong place to do the retry. The dist sender should not be returning an error if the first range isn't available on gossip. It should probably be retrying itself, not returning an error. That would necessitate that every client must do its own retries in case it runs into this same problem when connecting to a cluster where gossip doesn't have the first range descriptor.

tamird commented:
@spencerkimball doesn't that presume that `allocateNodeID` will need to move down the stack? at this level we don't know the cause of the error.

tschottdorf commented:
It's operating on a bare DistSender, and having something else down there that can cause infinite retries sounds like a pain. I'm happiest about retries at the top level, clients need retry logic anyways and they don't need to care about the specific reason, that's why we have `proto.Error`.

spencerkimball commented:
@tamird I don't understand your comment. What do you mean by "down the stack"? My point is that this level shouldn't be dealing with retryable errors. There are plenty of retryable error types, currently all handled in dist_sender. Why not continue handling those there?

spencerkimball commented:
@tschottdorf What do you mean a bare DistSender. This code is using a "db" client object. We also use the same "db" client object when doing a split or a merge. But I notice you haven't also made changes there to retry on retryable errors. Why not? The exact same error regarding lack of first range metadata in gossip might occur anywhere in the system which we use a "db" client. It's not reasonable to build retry loops for this kind of thing into all of them. If the dist sender can retry a retryable error, it should be retrying. Otherwise, why not just let all of the retryable errors propagate from the dist sender and wrap all of our client usage in retry loops?

tamird commented:
@spencerkimball ah, so you're suggesting that this retry goes here: https://github.com/cockroachdb/cockroach/blob/master/kv/dist_sender.go#L663 ?

tschottdorf commented:
@tamird @spencerkimball I think this slipped through the cracks here then: https://github.com/cockroachdb/cockroach/pull/1248/files#r31667485

tamird commented:
Yup, seems so. @tschottdorf want to try adding that retry there and seeing what the fallout is?

spencerkimball commented:
Yeah, we just weren't considering the possibility of lacking the first range descriptor. Let's just make sure that the resulting error is retryable and allow the result to trigger this retry loop.

tschottdorf commented:
I'll back out this change and do the proper retry instead. Not sure I'll get to it today though (it's already late here), so if either of you is burning to get this done, feel free to just go for it.

spencerkimball commented:
No rush!

On Monday, June 22, 2015, Tobias Schottdorf <notifications@github.com>
wrote:

> I'll back out this change and do the proper retry instead. Not sure I'll
> get to it today though (it's already late here), so if either of you is
> burning to get this done, feel free to just go for it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1323#issuecomment-114178717>
> .
>
